### DISCUSSION TITLE
How to dynamically modify commit author with transient

### DISCUSSION THREAD

**Discussion by genehack** [magit/magit#4371] [Support] [answered]
Formerly, I was able to use a `.dir-locals.el` file to dynamically modify the arguments Magit passed to the underlying Git commit command, like so:

```elisp
((nil . ((magit-commit-arguments . ("--author=John SJ Anderson <john.anderson@work>")))))
```

(This allows me to use different email addresses in different directories, so my work projects have commits from my work address, my personal projects my own email, open source projects use an email associated with the project, etc.)

However, with the switch to transient, this no longer works. I tried to read the transient manual and very rapidly got very confused. I understand how to set a value in the popup and save it; I no longer understand how to make it automatically changed based on something buffer-local. 

Can somebody please explain how I do this in the brave new `transient` Magit world? Thanks.

**Comment by kyleam** [comment, marked as answer]
[ I'm not that familiar with Transient, but I'm sure Jonas will correct me if I say something too off ]

I don't think you can map that to something similar for Transient.  The closest I can get is something like this:

```elisp
(require 'magit)
(require 'seq)

(defvar-local my/magit-commit-author nil)
(put 'my/magit-commit-author 'safe-local-variable 'stringp)

(defun my/magit-commit-arguments ()
  (if (eq transient-current-command 'magit-commit)
      (magit-commit-arguments)
    (let ((args (alist-get 'magit-commit transient-values)))
      (when my/magit-commit-author
        (setq args
              (cons (concat "--author=" my/magit-commit-author)
                    (seq-remove (lambda (x) (string-prefix-p "--author=" x))
                                args))))
      args)))

(oset (get 'magit-commit 'transient--prefix) :value
      #'my/magit-commit-arguments)

;; .dir-locals.el:
;; ((nil . ((my/magit-commit-author . "John SJ Anderson <john.anderson@work>"))))
```

But I think that probably reaches too far into the internals of Transient.

> (This allows me to use different email addresses in different directories, so my work projects have commits from my work address, my personal projects my own email, open source projects use an email associated with the project, etc.)

Hmm, is there a reason you're not configuring Git's `user.name` and `user.email` in the repository?  That has the advantage of working outside of Magit and also takes care of the committer name and email.

**Reply by genehack** [reply]
> Hmm, is there a reason you're not configuring Git's `user.name` and `user.email` in the repository? That has the advantage of working outside of Magit and also takes care of the committer name and email.

I do have those set at a global level, but want to use different ones in different repos, depending on where that repo lives in a directory hierarchy. 

On the command line, I solve for this by dynamically setting `GIT_COMMITTER_NAME` and `GIT_COMMITTER_EMAIL` to the appropriate values (via a system that hooks the `cd` command and sets or unsets variables based on a configuration). 

Configuring it at the repo level is _massively_ inconvenient; I have dozens of work repositories and need a more generalizable solution that doesn't require modifying the repo config in each of the many repo check outs I have.

**Reply by kyleam** [reply]
> I do have those set at a global level, but want to use different ones in different repos, depending on where that repo lives in a directory hierarchy. [...]
> Configuring it at the repo level is _massively_ inconvenient

You don't necessarily have to do it at the repo level.  If you have a directory hierarchy that reliably separates the identities, you can use conditional includes:

```
[includeIf "gitdir:/path/to/ident1-repo-top/"]
	   path = /path/to/ident1.conf
[includeIf "gitdir:/path/to/ident2-repo-top/"]
	   path = /path/to/ident2.conf
...
```

**Reply by genehack** [reply]
> > I do have those set at a global level, but want to use different ones in different repos, depending on where that repo lives in a directory hierarchy. [...]
> > Configuring it at the repo level is _massively_ inconvenient
> 
> You don't necessarily have to do it at the repo level. If you have a directory hierarchy that reliably separates the identities, you can use conditional includes:

Yep, got that set too -- see https://git.genehack.net/genehack/gitconfig/src/branch/main/etc/gitconfig#L5.

Commits made through Magit still end up with my base user info unless I manually set the `-A` option in the commit popup.

**Reply by kyleam** [reply]
> Commits made through Magit still end up with my base user info unless I manually set the `-A` option in the commit popup.

See gh-4360, but Magit should use the same identify as you'd see using git directly (and does on my end).

**Reply by genehack** [reply]
> See [gh-4360](https://github.com/magit/magit/discussions/4360), but Magit should use the same identify as you'd see using git directly (and does on my end).

Thanks! Reading through that ticket got me to look at my configs again, and I found a silly error on my part. After correcting that, things are working as desired. Appreciate your patience!
